File: Components/projects/Components_Development/CustomComponentsResources/web/js/custom/zk/components/Errmsg.js
//error message widget class
//this file will be loaded because
//the widget-class --  'custom.zk.components.Errmsg'
//specified in lang-addon.xml and zk.wpd
custom.zk.components.Errmsg = zk.$extends(zul.Widget, {
//	block for define attributes
//	where 'anAttr: null' in this block will
//	generate 3 items in this widget
//	1. this._anAttr
//	2. this.setAnAttr(value)
//	3. this.getAnAttr(value)
	
//	'anAttr: function (v)' will also generate the 3 items above,
//	the function will be called by setter,
//	i.e., setAnAttr(value) {
//		...
//		function(v)
//		...
//	}
//	so you can specify what you want to do while the setter is called

//	for more information, please refer to
//	http://www.zkoss.org/javadoc/latest/jsdoc/_global_/zk.html#$extends(zk.Class, _global_.Map, _global_.Map)
//	http://www.zkoss.org/javadoc/latest/jsdoc/_global_/zk.html#define(zk.Class, _global_.Map)
	$define: {
		msg: function (v) {
			var n;
			if (v
				&& (n = this.$n()))
				n.innerHTML = v;
		},
		description: null
	},
	// function to provide css class
	getZclass: function () {
		var zcls = this._zclass;
		return zcls? zcls : 'z-errmsg';
	}
});

File: Components/projects/Components_Development/src/custom/zk/components/Errmsg.java
package custom.zk.components;


import org.zkoss.zul.impl.XulElement;
// this file will be loaded because the
// <component-class>custom.zk.components.Errmsg</component-class>
// specified in lang-addon.xml
public class Errmsg extends XulElement {
	private String _msg;
	private String _description;

	public void setMsg (String msg) {
		_msg = msg;
		// this will call client side widget's setMsg(_msg)
		smartUpdate("msg", _msg);
	}
	public String getMsg () {
		return _msg;
	}
	public void setDescription (String description) {
		_description = description;
		// this will call client side widget's setDescription(_description)
		smartUpdate("description", _description);
	}
	public String getDescription () {
		return _description;
	}
	public String getZclass() {
		return _zclass == null ? "z-errmsg" : _zclass;
	}
	//-- ComponentCtrl --//
//	the renderProperties will be called by framework automatically,
//	remember to render super's properties first
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
	throws java.io.IOException {
		super.renderProperties(renderer);
		// this will call client side widget's setMsg(_msg)
		if (_msg != null)
			render(renderer, "msg", _msg);
		// this will call client side widget's setDescription(_description)
		if (_description != null)
			render(renderer, "description", _description);
	}
}
